managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/EntryImpl.java

```diff
@@ -22,36 +22,60 @@
 import io.netty.buffer.ByteBuf;
 import io.netty.buffer.RecyclableDuplicateByteBuf;
 import io.netty.buffer.Unpooled;
+import io.netty.util.Recycler;
+import io.netty.util.Recycler.Handle;
 
 final class EntryImpl implements Entry, Comparable<EntryImpl>, ReferenceCounted {
 
-    private final PositionImpl position;
-    private final ByteBuf data;
-
-    EntryImpl(LedgerEntry ledgerEntry) {
-        this.position = new PositionImpl(ledgerEntry.getLedgerId(), ledgerEntry.getEntryId());
-        this.data = ledgerEntry.getEntryBuffer();
+    private static final Recycler<EntryImpl> RECYCLER = new Recycler<EntryImpl>() {
+        @Override
+        protected EntryImpl newObject(Handle handle) {
+            return new EntryImpl(handle);
+        }
+    };
+    
+    private final Handle recyclerHandle;
+    private PositionImpl position;
+    private ByteBuf data;
+
+    public static EntryImpl create(LedgerEntry ledgerEntry) {
+        EntryImpl entry = RECYCLER.get();
+        entry.position = new PositionImpl(ledgerEntry.getLedgerId(), ledgerEntry.getEntryId());
```

Comment:
Regarding the life-cycle of the `PositionImpl`, we could return a new `PositionImpl` instance in `getPosition()` itself. The advantage would be that the object that is being retained for longer time (while in the cache) is going to get reused